✨ Set up comprehensive Copilot instructions with complete API design architecture and framework overview for Python 3.12 and CUDA 12 (#1031)
NVIDIA-NeMo/Curator: .github/copilot-instructions.md

### Common Development Patterns

#### Processing Pipeline Structure
```python
# Task-centric pipeline following API design patterns
from nemo_curator.core.client import RayClient
from nemo_curator.pipeline import Pipeline
from nemo_curator.stages.text.classifiers import FineWebMixtralEduClassifier
from nemo_curator.stages.text.io.reader.jsonl import JsonlReader
from nemo_curator.stages.text.io.writer.jsonl import JsonlWriter

# Example workflow following API design
ray_client = RayClient()
ray_client.start()

# Define pipeline stages
read_stage = JsonlReader(input_file_path, files_per_partition=1)
classifier_stage = FineWebMixtralEduClassifier()
write_stage = JsonlWriter(output_file_path)

# Build and run pipeline
pipeline = Pipeline(name="classifier_pipeline")
pipeline.add_stage(read_stage)
pipeline.add_stage(classifier_stage)
pipeline.add_stage(write_stage)

result = pipeline.run()
ray_client.stop()
```

#### GPU vs CPU Handling
```python
# Resource specification for stages
from nemo_curator.stages.resources import Resources

class MyProcessingStage(ProcessingStage):
    @property
    def resources(self) -> Resources:
        return Resources(cpus=1.0, gpu_memory_gb=4.0)
```

> 

Reviewer: @copilot , Please add more info from the api_design which looks like this.  Other things mostly looks good. 

```
# NeMo Curator Ray API Design

## Design Principles

### Task-Centric Architecture
Unlike the previous dataset-level operations, the new design operates on individual **Tasks** - batches of data that flow through the pipeline. This enables:
- Finer-grained control and monitoring
- Better resource utilization

### Map-style (Data-Parallel) Execution
All stages are designed to be map-style on tasks, meaning they take task as input and produce task as output. This allows for easy parallelization and scaling.
- We do not enforce 1-1 mapping between input and output tasks, but rather allow for multiple output tasks from a single input task and multiple input tasks from a single output task. More specifically, a stage applies a transformation from `X` to `Y`, where both `X` and `Y` can be `Task | list[Task] | None`.

### Fault Tolerance Requirements
**All stages MUST be fault-tolerant and retry-safe.** This is a critical requirement because:

- **Task Preemption:** Xenna can preempt/kill running tasks before completion and potentially reschedule them later, especially during autoscaling events
- **Partial Operations:** Tasks may be interrupted mid-execution, leaving partial state (e.g., incomplete file downloads)

## Core Components

1. **Tasks** : The fundamental unit of data that flows through the curation pipeline.
2. **Stages** : The fundamental processing unit that takes Tasks as input and produces Tasks as output.
3. **Pipelines** : A collection of stages that defines the complete processing workflow.
4. **Executors** : The component that orchestrates the execution of the pipeline.

### Tasks

A **Task** is the fundamental unit of data that flows through the curation pipeline, representing a batch of input data for processing.

#### Base Task Implementation

```python
@dataclass
class Task(ABC, Generic[T]):
    """Abstract base class for tasks in the pipeline."""
    task_id: str
    dataset_name: str
    data: T
    _stage_perf: list[StagePerfStats] = field(default_factory=list)
    _metadata: dict[str, Any] = field(default_factory=dict)

    @property
    @abstractmethod
    def num_items(self) -> int:
        """Get the number of items in this task."""

    @abstractmethod
    def validate(self) -> bool:
        """Validate the task data."""
```

#### Example Task Types

#### DocumentTask
For text-based curation pipelines:

```python
@dataclass
class DocumentBatch(Task[pa.Table | pd.DataFrame]):
    """Task for document processing."""

    @property
    def num_items(self) -> int:
        return len(self.data)

    def validate(self) -> bool:
        return isinstance(self.data, pd.DataFrame) and not self.data.empty
```

### Stages

#### Base Stage Interface

```python
class ProcessingStage(ABC, Generic[X, Y], metaclass=StageMeta):
    """Base class for all processing stages that accepts a task of type X and outputs a task of type Y."""

    @property
    @abstractmethod
    def name(self) -> str:
        """Unique name for this stage."""

    @property
    def resources(self) -> Resources:
        """Resource requirements for this stage."""
        return Resources(cpus=1.0)

    @abstractmethod
    def process(self, task: X) -> Y | list[Y]:
        """Process a single task, can output one or more task."""
```

#### Resource Specification

```python
@dataclass
class Resources:
    """Define resource requirements for a processing stage."""
    cpus: float = 1.0 # Number of CPU cores
    gpu_memory_gb: float = 0.0 # Number of GPU memory in GB (Only for single GPU)
    gpus: float = 0.0 # Number of GPUs (Only for multi-GPU)
    nvdecs: int = 0 # Number of NVDEC decoders
    nvencs: int = 0 # Number of NVENC encoders
    entire_gpu: bool = False # Whether to use the entire GPU
```

### Pipelines

A **Pipeline** is a collection of stages that defines the complete processing workflow.

```python
class Pipeline:
    """A pipeline defines a sequence of processing stages."""

    def __init__(self, stages: list[ProcessingStage]):
        self.stages = stages

    def add_stage(self, stage: ProcessingStage):
        """Add a stage to the pipeline."""

    def run(self, executor: BaseExecutor | None = None) -> list[Task] | None:
        """Run the pipeline."""
```

### Executors (Advanced)

**Executors** are responsible for running pipelines on different backends while maintaining a unified interface.
They do so with the help of **Adapters** which are the translation piece between our `ProcessingStage` and the desired "executor".
Each Executor runs a `list[ProcessingStage]` and then wraps each `ProcessingStage` to an `Adapter`, and then finally those wrapped classes, i.e adapters are executed.

#### Base Executor Interface

```python
class BaseExecutor(ABC):
    """Executor for a pipeline."""

    def __init__(self, config: dict[str, Any] | None = None):
        self.config = config or {}

    @abstractmethod
    def execute(self, stages: list[ProcessingStage], initial_tasks: list[Task] | None = None) -> None:
        """Execute the pipeline."""
```

### Backend Implementations

#### Xenna Executor
```python
class XennaExecutor(BaseExecutor):
    """Ray-based executor using Xenna backend."""

    def execute(self, stages: list[ProcessingStage], initial_tasks: list[Task] | None = None) -> None:
        # Convert stages to Xenna acceptable format using Xenna Adapters
        # Handle resource allocation
        # Execute with autoscaling
```

#### Ray Data Executor
```python
class RayDataExecutor(BaseExecutor):
    """Ray Data-based executor."""

    def execute(self, stages: list[ProcessingStage], initial_tasks: list[Task] | None = None) -> None:
        # Convert to Ray Data operations
        # Execute pipeline
```

## Development Environment

### Python Environment
- **Python Version**: 3.12 (recommended and tested), supports 3.10-3.12
- **Package Manager**: [uv](https://docs.astral.sh/uv/) for fast, reliable dependency management  
- **Virtual Environment**: Use uv's built-in virtual environment management
- **Note**: Python 3.11 is not supported due to RAPIDS compatibility requirements

### CUDA Environment (Optional)
- **CUDA Version**: 12.x (12.0+) for GPU acceleration
- **GPU Requirements**: NVIDIA GPU with Volta™ architecture or higher (compute capability 7.0+)
- **GPU Libraries**: RAPIDS (cuDF, cuML), PyTorch with CUDA 12 support, CuPy
- **Note**: GPU features are optional and CPU fallbacks are provided

## Key Technologies and Frameworks

### Core Dependencies
- **PyTorch**: Deep learning framework with CUDA support
- **Ray**: Distributed computing framework for data processing
- **Pandas/CuDF**: Data manipulation (CPU/GPU respectively)
- **Transformers**: Hugging Face transformers library
- **LoguRu**: Structured logging

### Modality-Specific Libraries

> - **Text Processing**: BeautifulSoup, fasttext, sentencepiece, trafilatura

Reviewer: Add Pytorch here.

- **Imports**: Follow import sorting conventions

### Module Structure
- **Stages**: Processing stages organized by modality (`text/`, `video/`, `audio/`, `image/`)
- **Utils**: Shared utilities for common operations
- **Datasets**: Data loading and manipulation classes
- **Modules**: Core processing algorithms

### Error Handling Patterns
```python
# Graceful degradation for optional GPU dependencies
try:
    import cudf
    import cvcuda
    import PyNvVideoCodec as Nvc
    HAS_GPU_SUPPORT = True
except ImportError:
    HAS_GPU_SUPPORT = False
    # Provide helpful error messages for missing dependencies
    logger.warning("GPU dependencies not available, falling back to CPU processing")

# Example from nvcodec_utils.py
try:
    import cvcuda
    import nvcv
    import pycuda.driver as cuda
    import PyNvVideoCodec as Nvc
    pixel_format_to_cvcuda_code = {
        Nvc.Pixel_Format.YUV444: cvcuda.ColorConversion.YUV2RGB,
        Nvc.Pixel_Format.NV12: cvcuda.ColorConversion.YUV2RGB_NV12,
    }
except (ImportError, RuntimeError):
    logger.warning("PyNvVideoCodec is not installed, some features will be disabled.")
    Nvc = None
    pixel_format_to_cvcuda_code = {}

> ```

Reviewer: We dont support this today.  Please remove fallback

# Run specific test categories
uv run pytest -m "not gpu"  # CPU tests only

# Run tests for specific modules
uv run pytest tests/stages/text/
uv run pytest tests/stages/image/
```

## Build and Development

### Development Workflow
```bash
# Install development dependencies
uv sync

# Run linting and formatting
uv run ruff check .
uv run ruff format .

# Run tests
uv run pytest

# Run specific test modules
uv run pytest tests/utils/test_nvcodec_utils.py

# Build documentation (if working on docs)
make docs-html

# Check for dependency updates
uv lock --upgrade
```

### Common Development Patterns

#### Processing Pipeline Structure
```python
# Standard pipeline component structure
class ProcessingStage:
    def __init__(self, config: Dict[str, Any]):
        self.config = config
    
    def __call__(self, dataset: DocumentDataset) -> DocumentDataset:
        # Process and return modified dataset
        pass
```

#### GPU vs CPU Handling
```python
# Check for GPU availability and fallback gracefully
if HAS_GPU_SUPPORT and use_gpu:
    # Use GPU-accelerated processing
    df = cudf.DataFrame(data)
else:
    # Fallback to CPU processing
    df = pd.DataFrame(data)

> ```

Reviewer: Look at api-design and follow that for instructions. 

https://github.com/NVIDIA-NeMo/Curator/blob/main/api-design.md

Another example workflow is:
```
"""
Example: Distributed Data Classification with NeMo Curator
FineWebMixtralEduClassifier Tutorial
"""

import os
import pandas as pd
from nemo_curator.core.client import RayClient
from nemo_curator.pipeline import Pipeline
from nemo_curator.stages.text.classifiers import FineWebMixtralEduClassifier
from nemo_curator.stages.text.io.reader.jsonl import JsonlReader
from nemo_curator.stages.text.io.writer.jsonl import JsonlWriter

# ---------------------------------------------------------
# Step 1: Silence Ray logs
# ---------------------------------------------------------
os.environ["LOGURU_LEVEL"] = "ERROR"

# ---------------------------------------------------------
# Step 2: Start Ray cluster
# ---------------------------------------------------------
ray_client = RayClient()
ray_client.start()

# ---------------------------------------------------------
# Step 3: Create sample input data
# ---------------------------------------------------------
input_file_path = "./input_data_dir"
texts = [
    "Quantum computing is set to revolutionize the field of cryptography.",
    "Investing in index funds is a popular strategy for long-term financial growth.",
    "Recent advancements in gene therapy offer new hope for treating genetic disorders.",
    "Online learning platforms have transformed the way students access educational resources.",
]

df = pd.DataFrame({"text": texts})
os.makedirs(input_file_path, exist_ok=True)
df.to_json(f"{input_file_path}/data.jsonl", orient="records", lines=True)

# ---------------------------------------------------------
# Step 4: Define pipeline stages
# ---------------------------------------------------------
read_stage = JsonlReader(input_file_path, files_per_partition=1)
classifier_stage = FineWebMixtralEduClassifier()
output_file_path = "./fineweb_mixtral_edu_classifier_results"
write_stage = JsonlWriter(output_file_path)

# ---------------------------------------------------------
# Step 5: Build and run pipeline
# ---------------------------------------------------------
classifier_pipeline = Pipeline(name="classifier_pipeline")
classifier_pipeline.add_stage(read_stage)
classifier_pipeline.add_stage(classifier_stage)
classifier_pipeline.add_stage(write_stage)

result = classifier_pipeline.run()

# ---------------------------------------------------------
# Step 6: Inspect results
# ---------------------------------------------------------
result_file = result[0].data[0]
result_df = pd.read_json(result_file, lines=True)
print(result_df.head())

# ---------------------------------------------------------
# Step 7: Stop Ray cluster
# ---------------------------------------------------------
ray_client.stop()

```

#### Configuration Loading
```python
# Use hierarchical configuration loading
from nemo_curator.utils.config_utils import load_config

config = load_config(
    config_path="config/processing.yaml",
    overrides={"batch_size": 1024}
)

> ```

Reviewer: We dont support this cleanly today

from nemo_curator.core.client import RayClient
from nemo_curator.pipeline import Pipeline
from nemo_curator.stages.text.classifiers import FineWebMixtralEduClassifier
from nemo_curator.stages.text.io.reader.jsonl import JsonlReader
from nemo_curator.stages.text.io.writer.jsonl import JsonlWriter

# Example workflow following API design
ray_client = RayClient()
ray_client.start()

# Define pipeline stages
read_stage = JsonlReader(input_file_path, files_per_partition=1)
classifier_stage = FineWebMixtralEduClassifier()
write_stage = JsonlWriter(output_file_path)

# Build and run pipeline
pipeline = Pipeline(name="classifier_pipeline")
pipeline.add_stage(read_stage)
pipeline.add_stage(classifier_stage)
pipeline.add_stage(write_stage)

result = pipeline.run()
ray_client.stop()
```

#### GPU vs CPU Handling
```python
# Resource specification for stages
from nemo_curator.stages.resources import Resources

class MyProcessingStage(ProcessingStage):
    @property
    def resources(self) -> Resources:
        return Resources(cpus=1.0, gpu_memory_gb=4.0)
```

## API Design Architecture

### Design Principles

#### Task-Centric Architecture

> Unlike the previous dataset-level operations, the new design operates on individual **Tasks** - batches of data that flow through the pipeline. This enables:

Reviewer: ```suggestion
The design of NeMo Curator is based on Ray and operates on individual **Tasks** - batches of data that flow through the pipeline. This enables:
```

Reviewer: @copilot , please address

# GitHub Copilot Instructions for NVIDIA NeMo Curator

## Overview

NVIDIA NeMo Curator is a generic, pluggable framework for building distributed data processing pipelines using Ray. It provides a unified API for running the same pipeline logic across different Ray orchestration backends.

### 🎯 Core Concept

> Enable users to define a data processing pipeline once and execute it using different Ray backends in a multi-node setting (Xenna, Ray Actors, Ray Data) without changing the pipeline logic

Reviewer 2: ```suggestion
Enable users to define a data processing pipeline once and execute it using different Ray backends in a multi-node setting (Xenna, Ray Actors, Ray Data) without changing the pipeline logic.
```

This scalable data preprocessing tool focuses on data curation pipelines for text, audio, video, and image modalities with support for both CPU and GPU processing.

## Development Environment

### Python Environment
- **Python Version**: 3.12 (recommended and tested), supports 3.10-3.12
- **Package Manager**: [uv](https://docs.astral.sh/uv/) for fast, reliable dependency management  
- **Virtual Environment**: Use uv's built-in virtual environment management

> - **Note**: Python 3.11 is not supported due to RAPIDS compatibility requirements

Reviewer 2: Still wondering if this is true?

Reviewer: Removed, good call out

> - **Note**: Python 3.11 is not supported due to RAPIDS compatibility requirements

Reviewer: ```suggestion
```

### CUDA Environment (Optional)
- **CUDA Version**: 12.x (12.0+) for GPU acceleration
- **GPU Requirements**: NVIDIA GPU with Volta™ architecture or higher (compute capability 7.0+)
- **GPU Libraries**: RAPIDS (cuDF, cuML), PyTorch with CUDA 12 support, CuPy
- **Note**: GPU features are optional and CPU fallbacks are provided

## Key Technologies and Frameworks

### Core Dependencies
- **PyTorch**: Deep learning framework with CUDA support
- **Ray**: Distributed computing framework for data processing
- **Pandas/CuDF**: Data manipulation (CPU/GPU respectively)
- **Transformers**: Hugging Face transformers library

> - **LoguRu**: Structured logging

Reviewer: @copilot , please address this review.

> - **LoguRu**: Structured logging

Reviewer: @copilot , please address this review.

    def num_items(self) -> int:
        return len(self.data)

    def validate(self) -> bool:
        return isinstance(self.data, pd.DataFrame) and not self.data.empty
```

### Stages

#### Base Stage Interface

```python
class ProcessingStage(ABC, Generic[X, Y], metaclass=StageMeta):
    """Base class for all processing stages that accepts a task of type X and outputs a task of type Y."""

    @property
    @abstractmethod
    def name(self) -> str:
        """Unique name for this stage."""

    @property
    def resources(self) -> Resources:
        """Resource requirements for this stage."""
        return Resources(cpus=1.0)

    @abstractmethod
    def process(self, task: X) -> Y | list[Y]:
        """Process a single task, can output one or more task."""
```

#### Resource Specification

```python
@dataclass
class Resources:
    """Define resource requirements for a processing stage."""
    cpus: float = 1.0 # Number of CPU cores
    gpu_memory_gb: float = 0.0 # Number of GPU memory in GB (Only for single GPU)
    gpus: float = 0.0 # Number of GPUs (Only for multi-GPU)
    nvdecs: int = 0 # Number of NVDEC decoders
    nvencs: int = 0 # Number of NVENC encoders

>     entire_gpu: bool = False # Whether to use the entire GPU

Reviewer: @copilot , please address this review.
